File: 73_set_matrix_zeroes/solution.py
from typing import List


class Solution:
  def setZeroes(self, matrix: List[List[int]]) -> None:
    ROWS, COLS = len(matrix), len(matrix[0])
    unique_rows, unique_cols = set(), set()

    """
    Finding the unique set of rows and cols
    where 0 is present
    """
    for i in range(ROWS):
      for j in range(COLS):
        if matrix[i][j] == 0:
          unique_rows.add(i)
          unique_cols.add(j)
    
    """
    Making all the rows as 0
    """
    for row in unique_rows:
      for col in range(COLS):
        matrix[row][col] = 0

    """
    Making all the cols as 0
    """
    for col in unique_cols:
      for row in range(ROWS):
        matrix[row][col] = 0


matrix = [[2, 4, 3], [1, 0, 0]]
n, m = 2, 3
solution_obj = Solution()
solution_obj.setZeroes(matrix)
print(matrix)

File: 73_set_matrix_zeroes/solution.c
#include <stdio.h>

int isPresent(int arr[], int x, int size)
{
    for (int i = 0; i < size; i++)
        if (arr[i] == x)
            return 1;
    return 0;
}

void setZeroes(int **matrix, int matrixSize, int *matrixColSize)
{
    int ROWS = matrixSize;
    int COLS = *matrixColSize;
    int unique_rows[200], row_idx = 0;
    int unique_cols[200], col_idx = 0;
    int i, j;

    for (i = 0; i < 200; i++)
    {
        unique_rows[i] = -1;
        unique_cols[i] = -1;
    }

    for (i = 0; i < ROWS; i++)
    {
        for (j = 0; j < COLS; j++)
        {
            if (matrix[i][j] == 0)
            {
                if (isPresent(unique_rows, i, 200) == 0)
                    unique_rows[row_idx++] = i;
                if (isPresent(unique_cols, j, 200) == 0)
                    unique_cols[col_idx++] = j;
            }
        }
    }

    for (i = 0; i < row_idx; i++)
    {
        for (j = 0; j < COLS; j++)
            matrix[unique_rows[i]][j] = 0;
    }

    for (j = 0; j < col_idx; j++)
    {
        for (i = 0; i < ROWS; i++)
            matrix[i][unique_cols[j]] = 0;
    }
}

int main()
{
    int matrix[][4] = {{0, 1, 2, 0}, {3, 4, 5, 2}, {1, 3, 1, 5}};

    displayMatrix(matrix, 3, 4);

    return 0;
}